Collect (sparse) matrix operations in a single module

keileg opened the pull request:
This PR collects functions related to operations on sparse matrices into a single module, accessible as `pp.matrix_operations`. The functions where previously located in `pp.utils.matrix_compression` (now deleted), `pp.fvutils`, `pp.utils.sparse_mat`. 

There are no functional changes in this PR, but several other files and tests had to updated to comply with the new file.

Also did random maintenance, including more typing and removal of some logging statements.

This PR goes some way towards implenting #476, although more work is needed to close that one.

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -35,23 +40,34 @@ def zero_columns(A, cols):
-def zero_rows(A, rows):
+def zero_rows(
+    A: sps.csr_matrix, rows: np.ndarray, diag: Optional[np.ndarray] = None
+) -> None:
```
See previous comment about diag. Consider rename to diag_values? Also, using "diag" assumes A is square, right? All in all, do we really want this?

keileg replied:
Agree, but upon looking, I also found this is unused code, so I removed the option instead.

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -81,7 +102,6 @@ def merge_matrices(A, B, lines):
     ------
```
Wouldn't it be safer to have two functions, replace_rows and replace_columns? I am extremely wary of assumptions on matrix format.

keileg replied:
Possibly, this is an old function that I just copied. I'll see what I can do with a reasonable effort.

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -488,3 +502,332 @@ def _csx_matrix_from_blocks(
+
+        Parameters
+        ----------
+        A sps.crs-matrix, to be inverted
```
```suggestion
        a: sps.crs-matrix, to be inverted
```

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -488,3 +502,332 @@ def _csx_matrix_from_blocks(
+        Parameters
+        ----------
+        A sps.crs-matrix, to be inverted
+        sz - size of the individual blocks
```
```suggestion
        sz: size of the individual blocks
```

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -488,3 +502,332 @@ def _csx_matrix_from_blocks(
+            in itself is not efficient, but it is hopefully well suited for
+            speeding up with numba.
+
+            It may be possible to restruct the code to further help numba,
```
```suggestion
            It may be possible to restructure the code to further help numba,
```

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -488,3 +502,332 @@ def _csx_matrix_from_blocks(
+def rlencode(A: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    """Compress matrix by looking for identical columns.
+
+    Example usage: Convert the a full set of (row or column) indices of a
```
```suggestion
    Example usage: Convert a full set of (row or column) indices of a
```

IvarStefansson commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -488,3 +502,332 @@ def _csx_matrix_from_blocks(
+            be repeated to restore the original array.
+
+    See also:
+        rlencode
```
```suggestion
        rldecode
```

IvarStefansson commented on src/porepy/utils/graph.py:
```diff
@@ -65,6 +61,8 @@ def bfs(self, start, color):
             if node not in visited:
                 visited.append(node)
-                neighbours = sparse_mat.slice_indices(self.node_connections, node)
+                neighbours = pp.matrix_operations.slice_indices(
```
Consider changing to neighbors to be consistent with rest of PorePy (US grammar). Mixing has lead to cumbersome debugging in the past!

keileg replied:
Fair point. Will do.

IvarStefansson commented on tests/integration/test_mpfa_mpsa_partial_update.py:
```diff
@@ -737,6 +737,1036 @@ def _test_one_cell_a_time_node_keyword(self):
+
+if __name__ == "__main__":
+    unittest.main()
+""" Unit tests for the partial update features of mpfa and mpsa.
```
This looks really strange. Have you accidentally pasted something here?

keileg replied:
Yes, you're probably right. Will fix.

IvarStefansson reviewed (commented):
A few questions to be addressed, some of them very minor

keileg commented on src/porepy/numerics/linalg/matrix_operations.py:
```diff
@@ -1,26 +1,31 @@
     Function to zero out columns in matrix A. Note that this function does not
     change the sparcity structure of the matrix, it only changes the column
-    values to 0
+    values to 0.
```
Agree, but upon looking, I also found this is unused code, so I removed the option instead.

keileg commented:
@IvarStefansson I think I have taken care of all comments. I removed the diagonal row option from methods to zero out rows and columns, so I think the main change is the way I fixed `merge_matrices`.
